### PR TITLE
A Rust rewrite based on IDesktopWallpaper Instead Of IActiveDesktop (#1)

### LINKED ISSUE #7: Rewrite in Rust
I think Rust would be much more maintainable than C++.

I asked ChatGPT to convert the existing C++ to Rust. That may be a good starting point:

```rust
extern crate com;

use com::{com_interface, ComRc, ComItf};
use std::ffi::OsString;
use std::io::{self, Write};
use winapi::{LPWSTR, MAX_PATH, WPSTYLE_SPAN, WPSTYLE_CENTER, WPSTYLE_TILE, WPSTYLE_STRETCH, WPSTYLE_KEEPASPECT};

#[com_interface("6D5140C1-7436-11CE-8034-00AA006009FA")]
pub trait IActiveDesktop: ComItf {
fn GetWallpaper(&self, wp: LPWSTR, len: i32, flags: i32) -> HRESULT;
fn SetWallpaper(&self, wp: LPWSTR, flags: i32) -> HRESULT;
fn SetWallpaperOptions(&self, opts: &WALLPAPEROPT, flags: i32) -> HRESULT;
fn ApplyChanges(&self, flags: i32) -> HRESULT;
}

#[repr(C)]
#[derive(Debug, Copy, Clone)]
struct WALLPAPEROPT {
dwSize: u32,
dwStyle: i32,
}

fn get_wallpaper(sz_wallpaper: &mut [u16]) -> io::Result<()> {
unsafe {
let result = CoInitialize(std::ptr::null_mut());
if result != S_OK {
return Err(io::Error::new(
io::ErrorKind::Other,
"Failed to initialize COM",
));
}
let mut p_desk: *mut IActiveDesktop = std::ptr::null_mut();
let hr = CoCreateInstance(
&CLSID_ActiveDesktop,
std::ptr::null_mut(),
CLSCTX_INPROC_SERVER,
&IID_IActiveDesktop,
&mut p_desk as *mut *mut IActiveDesktop as *mut *mut std::ffi::c_void,
);
if hr != S_OK {
return Err(io::Error::new(
io::ErrorKind::Other,
"Failed to get the desktop wallpaper",
));
}
(*p_desk).GetWallpaper(sz_wallpaper.as_mut_ptr(), MAX_PATH, 0);
(*p_desk).Release();
CoFreeUnusedLibraries();
CoUninitialize();
}
Ok(())
}

fn set_wallpaper(full_path: &[u16], wp_style: i32) -> io::Result<()> {
unsafe {
let result = CoInitialize(std::ptr::null_mut());
if result != S_OK {
return Err(io::Error::new(
io::ErrorKind::Other,
"Failed to initialize COM",
));
}
let mut p_desk: *mut IActiveDesktop = std::ptr::null_mut();
let hr = CoCreateInstance(
&CLSID_ActiveDesktop,
std::ptr::null_mut(),
```

*(It's a little bit truncated)*

### PR DESCRIPTION
**Pull request by a-isaiahharvey**
Fixes #7

This rewrite uses the `windows` crate for using Win32 APIs and `clap` for creating the command line interface

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for working on this 🙏

// @beyluta

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+impl DesktopWallpaper {
+    pub fn new() -> Result<Self> {
+        let interface: IDesktopWallpaper;
+
+        unsafe {
+            CoInitialize(None)?;
+            interface = CoCreateInstance(&DesktopWallpaper, None, CLSCTX_LOCAL_SERVER)?;
+        }
```
Comment (sindresorhus) [inline src/lib.rs:60]:
```suggestion
        let interface: IDesktopWallpaper = unsafe {
            CoInitialize(None)?;
            CoCreateInstance(&DesktopWallpaper, None, CLSCTX_LOCAL_SERVER)?;
        }
```

?

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    ) -> Result<()> {
+        unsafe {
+            let full_path = format!(
+                "{}",
+                path.canonicalize()
+                    .expect("Could not canonicalize")
+                    .display()
+            );
+
+            let mut encoded = full_path.encode_utf16().chain([0u16]).collect::<Vec<u16>>();
```
Comment (sindresorhus) [inline src/lib.rs:121]: Is this really the best way to convert a Path to PCWSTR?

Maybe: https://stackoverflow.com/a/74173701/64949

Comment (sindresorhus) [reply src/lib.rs:121]: Does this part need to be inside `unsafe`?

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            let wallpaper_string = match wallpaper {
+                Ok(pwstr) => pwstr.to_string().ok()?,
+                Err(_) => String::new(),
+            };
+
+            let path = Path::new(&wallpaper_string);
+
+            if path.exists() && path.is_file() {
+                if let Some(path) = path.parent() {
+                    if path.is_dir() {
```
Comment (sindresorhus) [inline src/lib.rs:97]: Can the if statements here be chained instead of nested?

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                        monitor_index,
+                        wallpaper,
+                    })
+                }
+            })
+            .collect()
+    }
+
+    pub fn get_wallpaper(&self, monitor: &Monitor) -> Option<PathBuf> {
+        unsafe {
```
Comment (sindresorhus) [inline src/lib.rs:84]: Can you reduce the `unsafe` usage to only the code that absolutely have to be inside the `unsafe` block?

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            let wallpaper: Result<PWSTR> =
+                self.interface.GetWallpaper(PCWSTR(monitor.monitor_index.0));
+
+            let wallpaper_string = match wallpaper {
+                Ok(pwstr) => pwstr.to_string().ok()?,
+                Err(_) => String::new(),
```
Comment (sindresorhus) [inline src/lib.rs:90]: Empty string is not a good fallback. Propagate the error instead.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: src/main.rs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    #[command(subcommand)]
+    command: Commands,
+}
+
+#[derive(Debug, Subcommand)]
+enum Commands {
+    /// Gets current wallpaper
+    Get {
+        /// Index of monitor starting from 0
+        #[arg(short, long, default_value_t = 0)]
```
Comment (sindresorhus) [inline src/main.rs:19]: I was thinking maybe a `value_parser`. See: https://docs.rs/clap/latest/clap/_derive/_tutorial/index.html#validated-values

File: src/main.rs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            return;
+        }
+    };
+    let monitors = match wallpaper.get_monitors() {
+        Ok(monitors) => monitors,
+        Err(error) => {
+            eprintln!("Failed to retrieve available monitors: {error}");
+            return;
+        }
+    };
```
Comment (sindresorhus) [inline src/main.rs:68]: Couldn't you use `map_err` for cases like this when the value is just passed through?

File: src/main.rs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+fn main() {
+    let wallpaper = match DesktopWallpaper::new() {
+        Ok(wallpaper) => wallpaper,
+        Err(error) => {
+            eprintln!("{error}");
+            return;
+        }
+    };
```
Comment (sindresorhus) [inline src/main.rs:61]: I'm not familiar with `clap`. Will this cause the binary to have a non-zero exit code? If not, we should make sure all error have non-zero exit.

File: src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    pub fn get_wallpaper(&self, monitor: &Monitor) -> Option<PathBuf> {
+        let wallpaper: Result<PWSTR> =
+            unsafe { self.interface.GetWallpaper(PCWSTR(monitor.monitor_index.0)) };
+
+        let wallpaper_string = match wallpaper {
+            Ok(pwstr) => unsafe { pwstr.to_string().ok()? },
+            Err(error) => {
+                eprintln!("{error}");
```
Comment (sindresorhus) [inline src/lib.rs:86]: This part of the code is the library part, so for the sake of separation, I don't think it should be printing. Instead, I would make `get_wallpaper` return a Result.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: src/main.rs [hunk 10]
```diff
@@ -0,0 +1,89 @@
+use std::path::{Path, PathBuf};
+
+use clap::{command, Parser, Subcommand};
+use wallpaper::{DesktopWallpaper, DesktopWallpaperPosition, Monitor};
+
+/// Manage the desktop wallpaper on Windows
+#[derive(Debug, Parser)]
+#[command(author, version, about, long_about = None)]
```
[the remark this answers is not in the document]
Comment (Turbo87) [reply src/main.rs:8]: Example: https://github.com/rust-lang/crates.io/blob/722035b655764203ed42cd1069124144048f7d5c/src/bin/crates-admin.rs#L8-L9

File: src/main.rs (hunk 8 above)
Comment (connor4312) [reply src/main.rs:61]: As-written the program will always have an exit code of 0.

File: src/main.rs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+fn if_chosen_monitor_within_range(choice: usize, monitors: &[Monitor], func: impl Fn()) {
+    if choice < monitors.len() {
+        func();
+    } else {
+        eprintln!(
+            "The available monitors are from 0 to {} but {choice} was given",
+            monitors.len() - 1,
+        );
+    }
+}
```
Comment (connor4312) [inline src/main.rs:52]:
```suggestion
fn if_chosen_monitor_within_range(choice: usize, monitors: &[Monitor], func: impl FnOnce(&Monitor)) {
    match monitors.get(choice) {
        Some(m) => func(m),
        None => eprintln!(
            "The available monitors are from 0 to {} but {choice} was given",
            monitors.len() - 1,
        ),
    }
}
```

avoids extra bounds checking when you access the index

**Comment by sindresorhus** [comment]
@a-isaiahharvey Some more feedback here: https://mastodon.social/@thejpster@hachyderm.io/109797282634092665

File: src/main.rs (hunk 6 above)
Comment (sindresorhus) [reply src/main.rs:13]: I'm still a Rust noob, so I cannot help with that. Let's just add a TODO comment about it.

File: Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "wallpaper"
+version = "2.0.0"
+edition = "2021"
+authors = ["Sindre Sorhus"]
```
Comment (sindresorhus) [inline Cargo.toml:5]: @a-isaiahharvey You should add your name here. Only if you want, of course.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 12 above)
Comment (a-isaiahharvey) [reply Cargo.toml]: Thank you for the opportunity to be a co-author of one of your projects 😊

**Comment by sindresorhus** [comment]
@a-isaiahharvey Great work!

**Comment by sindresorhus** [comment]
@a-isaiahharvey Would you be able to compile the binary for me? Based on `main` branch: https://github.com/sindresorhus/win-wallpaper/commit/b55c6f6c20d629627d215d43836f9e8eed2547c2 (3.0.0)

**Comment by a-isaiahharvey** [comment]
[wallpaper-aarch64-pc-windows-msvc.zip](https://github.com/sindresorhus/win-wallpaper/files/10715591/wallpaper-aarch64-pc-windows-msvc.zip)
[wallpaper-x86_64-pc-windows-msvc.zip](https://github.com/sindresorhus/win-wallpaper/files/10715592/wallpaper-x86_64-pc-windows-msvc.zip)

**Comment by a-isaiahharvey** [comment]
@sindresorhus Here are the compiled binaries 
[wallpaper-aarch64-pc-windows-msvc.zip](https://github.com/sindresorhus/win-wallpaper/files/10715599/wallpaper-aarch64-pc-windows-msvc.zip)
[wallpaper-x86_64-pc-windows-msvc.zip](https://github.com/sindresorhus/win-wallpaper/files/10715600/wallpaper-x86_64-pc-windows-msvc.zip)

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/win-wallpaper/releases/tag/v3.0.0

**Pull request merged**